src/Agent.Worker/TaskManager.cs

```diff
@@ -120,12 +145,36 @@ private async Task DownloadAsync(IExecutionContext executionContext, Pipelines.T
             // first check to see if we already have the task
             string destDirectory = GetDirectory(task);
             Trace.Info($"Ensuring task exists: ID '{task.Id}', version '{task.Version}', name '{task.Name}', directory '{destDirectory}'.");
-            if (File.Exists(destDirectory + ".completed"))
+
+            var configurationStore = HostContext.GetService<IConfigurationStore>();
+            AgentSettings settings = configurationStore.GetSettings();
+            Boolean signingEnabled = !String.IsNullOrEmpty(settings.Fingerprint);
+
+            if (File.Exists(destDirectory + ".completed") && !signingEnabled)
             {
                 executionContext.Debug($"Task '{task.Name}' already downloaded at '{destDirectory}'.");
                 return;
             }
 
+            String taskZipPath = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.TaskZips), $"{task.Name}_{task.Id}_{task.Version}.zip"); // TODO: Use shared code to create this string.
+            if (signingEnabled && File.Exists(taskZipPath))
+            {
+                executionContext.Debug($"Task '{task.Name}' already downloaded at '{taskZipPath}'.");
+
+                // We need to extract the zip now because the task.json metadata for the task is used in JobExtension.InitializeJob.
+                // This is fine because we overwrite the contents at task run time.
+                if (!File.Exists(destDirectory + ".completed"))
+                {
+                    // The zip exists but it hasn't been extracted yet.
+                    IOUtil.DeleteDirectory(destDirectory, CancellationToken.None);
```

Comment:
consider putting these 4 lines in a method and reusing them in both places (here and line131)